File: src/it/scala/com/exasol/cloudetl/kafka/KafkaTopicDataImporterNullValueIT.scala
package com.exasol.cloudetl.kafka

import java.lang.{Integer => JInt}
import java.lang.{Long => JLong}

import scala.jdk.CollectionConverters._

import com.exasol.ExaDataTypeException
import com.exasol.ExaMetadata
import com.exasol.cloudetl.kafka.KafkaTopicDataImporterAvroIT.schemaRegistryUrl

import org.apache.kafka.common.TopicPartition
import org.apache.kafka.common.serialization.Serializer
import org.apache.kafka.common.serialization.StringSerializer
import org.apache.kafka.clients.producer.ProducerRecord
import org.mockito.ArgumentMatchers._
import org.mockito.Mockito.{times, verify, when}

class KafkaTopicDataImporterNullValueIT extends KafkaIntegrationTest {
  override def additionalProperties: Map[String, String] =
    Map("SCHEMA_REGISTRY_URL" -> schemaRegistryUrl)

  implicit val serializer: Serializer[String] = {
    val properties = Map("schema.registry.url" -> schemaRegistryUrl)
    val serializer = new StringSerializer()
    serializer.configure(properties.asJava, false)
    serializer
  }

  test("emit fills columns with null if record value is null") {
    createCustomTopic(topic)
    implicit val serializer = new StringSerializer
    publishToKafka(new ProducerRecord(topic, null: String))
    publishToKafka(new ProducerRecord(topic, null: String))

    val iter = mockExasolIterator(properties, Seq(0), Seq(-1))
    val meta = mock[ExaMetadata]
    when(meta.getOutputColumnCount()).thenReturn(4)
    KafkaTopicDataImporter.run(meta, iter)

    verify(iter, times(2)).emit(Seq(any[Object]): _*)
    verify(iter, times(1)).emit(null, null, JInt.valueOf(0), JLong.valueOf(0))
    verify(iter, times(1)).emit(null, null, JInt.valueOf(0), JLong.valueOf(1))
  }

}

File: src/test/java/com/exasol/cloudetl/kafka/KafkaTopicDataImporterNullValueIT.java
package com.exasol.cloudetl.kafka;

import static org.junit.jupiter.api.Assertions.assertAll;
import static org.mockito.ArgumentMatchers.*;
import static org.mockito.Mockito.*;

import java.util.*;

import org.apache.kafka.clients.producer.ProducerRecord;
import org.apache.kafka.common.serialization.StringSerializer;
import org.junit.jupiter.api.Test;

class KafkaTopicDataImporterNullValueIT extends KafkaIntegrationTest {
    @Override
    Map<String, String> additionalProperties() {
        return Map.of("SCHEMA_REGISTRY_URL", SCHEMA_REGISTRY_URL);
    }

    @Test
    void emitFillsColumnsWithNullIfRecordValueIsNull() throws Exception {
        createCustomTopic(this.topic);
        publishToKafka(new ProducerRecord<String, String>(this.topic, null), new StringSerializer());
        publishToKafka(new ProducerRecord<String, String>(this.topic, null), new StringSerializer());

        final var iterator = mockExasolIterator(this.properties, List.of(0), List.of(-1L));

        KafkaTopicDataImporter.run(KafkaTopicDataImporterJsonToColumnsIT.mockMetadata(Object.class, Object.class, Integer.class, Long.class),
                iterator);

        assertAll(() -> verify(iterator, times(2)).emit(any(Object[].class)),
                () -> verify(iterator).emit(null, null, 0, 0L),
                () -> verify(iterator).emit(null, null, 0, 1L));
    }
}
